pkg/monitortests/metrics/rate_analyzer.go

```diff
@@ -0,0 +1,55 @@
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	prometheustypes "github.com/prometheus/common/model"
+)
+
+// RateSeriesAnalyzer analyzes a rate bases timeseries
+type RateSeriesAnalyzer struct{}
+
+func (RateSeriesAnalyzer) Analyze(ctx context.Context, query QueryRunner, start, end time.Time, callback Callback) error {
+	result, err := query.RunQuery(ctx, start, end)
+	if err != nil {
+		return fmt.Errorf("query returned error, monitor: %s, err: %w", callback.Name(), err)
+	}
+	if result.Type() != prometheustypes.ValMatrix {
+		return fmt.Errorf("expected a prometheus Matrix type, monitor: %s", callback.Name())
+	}
+
+	zero := prometheustypes.SampleValue(0)
+	matrix := result.(prometheustypes.Matrix)
+	for _, series := range matrix {
+		func() {
+			callback.StartSeries(series.Metric)
+			defer callback.EndSeries()
+
+			var from, to *prometheustypes.SamplePair
+			for _, current := range series.Values {
```

Comment:
does the `range` creates the `current` var once and then reuses it ?
if yes, then we cannot simply store `&current` as it will point to the same variable.